modules/tinymce/src/core/main/ts/fmt/FormatUtils.ts

```diff
@@ -18,39 +18,71 @@ const isNode = (node: any): node is Node => !!(node).nodeType;
 const isElementNode = (node: Node): node is Element =>
   NodeType.isElement(node) && !Bookmarks.isBookmarkNode(node) && !isCaretNode(node) && !NodeType.isBogus(node);
 
-const isInlineBlock = (node: Node): boolean => {
-  return node && /^(IMG)$/.test(node.nodeName);
+// In TinyMCE, directly selected elements are indicated with the data-mce-selected attribute
+// Elements that can be directly selected include control elements such as img, media elements, noneditable elements and others
+const isElementDirectlySelected = (dom: DOMUtils, node: Node): boolean => {
+  // Table cells are a special case and are separately handled from native editor selection
+  if (isElementNode(node) && !/^(TD|TH)$/.test(node.nodeName)) {
+    const selectedAttr = dom.getAttrib(node, 'data-mce-selected');
+    const value = parseInt(selectedAttr, 10);
+    // Avoid cases where data-mce-selected is not a positive number e.g. inline-boundary
+    return !isNaN(value) && value > 0;
+  } else {
+    return false;
+  }
 };
 
 const isEditable = (elm: HTMLElement): boolean =>
   elm.isContentEditable === true;
 
-const moveStart = (dom: DOMUtils, selection: EditorSelection, rng: Range): void => {
-  const offset = rng.startOffset;
-  let container = rng.startContainer;
+const adjustSelectionAfter = (editor: Editor, action: () => void, shouldMoveStart: (startNode: Node) => boolean): void => {
+  const { selection, dom } = editor;
+  const selectedNodeBeforeAction = selection.getNode();
+  const isSelectedBeforeNodeNoneditable = NodeType.isContentEditableFalse(selectedNodeBeforeAction);
 
-  if (container === rng.endContainer) {
-    if (isInlineBlock(container.childNodes[offset])) {
-      return;
-    }
+  action();
+
+  // Check previous selected node before the action still exists in the DOM
+  // and is still noneditable
+  const isBeforeNodeStillNoneditable = isSelectedBeforeNodeNoneditable && NodeType.isContentEditableFalse(selectedNodeBeforeAction);
+  if (isBeforeNodeStillNoneditable && dom.isChildOf(selectedNodeBeforeAction, editor.getBody())) {
+    editor.selection.select(selectedNodeBeforeAction);
+  } else if (shouldMoveStart(selection.getStart())) {
+    moveStartToNearestText(dom, selection);
   }
+};
+
+// Note: The reason why we only care about moving the start is because MatchFormat and its function use the start of the selection to determine if a selection has a given format or not
+const moveStartToNearestText = (dom: DOMUtils, selection: EditorSelection): void => {
+  const rng = selection.getRng();
+  const { startContainer, startOffset } = rng;
+  const selectedNode = selection.getNode();
 
-  // Move startContainer/startOffset in to a suitable node
-  if (NodeType.isElement(container)) {
-    const nodes = container.childNodes;
+  if (isElementDirectlySelected(dom, selectedNode)) {
```

Comment:
This code path appears to be new, but doesn't look like it's tested anywhere for elements that aren't cef. Is that accurate?

Reply:
Yep, I am only testing CEF at the moment. I should be able and another one for formatting an image as another test case.

Reply:
Additional test has been added 👍